File: Data Structures & Algorithms/anagram-groups/submission-0.go
func groupAnagrams(strs []string) [][]string {
	sol := [][]string{}

	for _, cur := range strs {
		process(cur, &sol)
	}

	return sol
}

func process(str string, sol *[][]string) {
	for i, entries := range *sol {
		entry := entries[0]

		if compare(entry, str) {
			(*sol)[i] = append(entries,str)
			return 
		}
	}

	*sol = append(*sol, []string{str})
}

func compare(a, b string) bool {
	if len(a) != len(b) {
		return false
	}

	var counts [26]int

	for i, _ := range a {
		counts[a[i] - 'a']++
		counts[b[i] - 'a']--
	}

	for _, val := range counts {
		if val != 0 {
			return false
		}
	}

	return true
}

File: Data Structures & Algorithms/string-encode-and-decode/submission-0.py
class Solution:

    def encode(self, strs: List[str]) -> str:
        out = ""
        for s in strs:
            l = len(s)
            f = f"{l:03d}"
            out = out+ f + s
        return out


    def decode(self, s: str) -> List[str]:
        sol = []
        while s:
            l = int(s[:3])
            sol.append(s[3:3+l])
            s = s[3+l:]
        return sol


